service/tablet_allocator.cc

```diff
@@ -500,24 +502,30 @@ class load_balancer {
         std::vector<table_id_and_size_desc> tables_need_resize;
         std::vector<table_id_and_size_desc> tables_being_resized;
 
-        static bool table_needs_merge(const table_size_desc& d) {
-            // The initial_tablet_count is respected while the table is in "growing mode".
-            // We say that a table leaves this mode if it required a split above the initial
-            // tablet count. After that, we can rely purely on the average size to say that
-            // a table is shrinking and requires merge.
-            // FIXME: this is not perfect and we may want to leave the mode too if we detect
-            //  average size is decreasing significantly, before any split happened.
-            bool left_growing_mode = !d.resize_decision.initial_decision();
-            lblogger.debug("table_needs_merge: tablet_count={}, avg_tablet_size={}, left_growing_mode={} (seq number: {})",
-                           d.tablet_count, d.avg_tablet_size, left_growing_mode, d.resize_decision.sequence_number);
-            return left_growing_mode && d.tablet_count > 1 && d.avg_tablet_size < d.target_min_tablet_size();
-        }
-        static bool table_needs_split(const table_size_desc& d) {
-            return d.avg_tablet_size > d.target_max_tablet_size();
+        static locator::resize_decision to_resize_decision(const table_size_desc& d) {
+            locator::resize_decision decision;
+
+            auto target_tablet_count = d.tablet_count;
+            if (d.tablet_count < d.min_tablet_count || d.avg_tablet_size > d.target_max_tablet_size()) {
+                // Split based on min_tablet_count or avg_tablet_size
+                // TODO: extend to n-way split when needed
+                target_tablet_count *= 2;
+                decision.way = resize_decision::split{};
+            } else if (d.avg_tablet_size < d.target_min_tablet_size()) {
```

Comment:
I think @raphaelsc is right that we lose hysteresis here. If we emitted "split" decision because `d.avg_tablet_size` crossed `d.target_max_tablet_size()`, we shouldn't emit none when it moves back below `d.target_max_tablet_size()`, but when it moves back below `d.target_tablet_size()`. So the logic would be:

```
if (d.tablet_count < d.min_tablet_count || d.avg_tablet_size > d.target_max_tablet_size()
  || (d.current_request == split && d.avg_tablet_size > d.target_tablet_size)) {
     decision = split;
}
```

And similar for merge.

Reply:
Added hysteresis back in v16 ([fea93a8](https://github.com/scylladb/scylladb/commit/fea93a8715456fb8eaa30f6fb869337b40d29bb5))